Add `--retry-until-up`, `--region`, `--zone`, and `--idle-minutes-to-autostop` for interactive nodes

ewzeng opened the pull request:
Add `--retry-until-up`, `--region`, and `--idle-minutes-to-autostop` for interactive nodes. This fixes #1131 and #1205.

Tested: `sky {cpunode, gpunode}` on Azure:
- [x] With and without `--idle-minutes-to-autostop`
- [x] With varying regions (`--region`)

I also used print statements to verify `_launch_with_confirm` is called with the correct `retry_until_up` argument. Not sure how else to test that on Azure.

michaelzhiluo commented on sky/cli.py:
```diff
@@ -207,15 +207,36 @@ def _interactive_node_cli_command(cli_func):
+                                  is_flag=True,
+                                  default=False,
+                                  required=False,
+                                  help='Retry until up.')
```
Better description, such as:
```
@click.option(
    '--retry-until-up',
    '-r',
    default=False,
    is_flag=True,
    required=False,
    help=('Whether to retry provisioning infinitely until the cluster is up, '
          'if we fail to launch the cluster on any possible region/cloud due '
          'to unavailability errors.'))
```

ewzeng replied:
Sounds good.

michaelzhiluo commented on sky/cli.py:
```diff
@@ -207,15 +207,36 @@ def _interactive_node_cli_command(cli_func):
+                                 default=None,
+                                 type=str,
+                                 required=False,
+                                 help='The region to use.')
```
By a similar token:
`The region to use. If specified, overrides the default region....` (what is default region for AWS, GCP, ...)

ewzeng replied:
I don't think there's a default region - if `--region` is not specified SkyPilot will try all regions.

michaelzhiluo replied:
that's right, ignore this comment

michaelzhiluo commented on sky/cli.py:
```diff
@@ -207,15 +207,36 @@ def _interactive_node_cli_command(cli_func):
+                                 type=int,
+                                 required=False,
+                                 help=('Automatically stop the cluster after'
+                                       'this many minutes of idleness.'))
```
```
@click.option(
    '--idle-minutes-to-autostop',
    '-i',
    default=None,
    type=int,
    required=False,
    help=('Automatically stop the cluster after this many minutes '
          'of idleness, i.e., no running or pending jobs in the cluster\'s job '
          'queue. Idleness starts counting after setup/file_mounts are done; '
          'the clock gets reset whenever there are running/pending jobs in the '
          'job queue. '
          'Setting this flag is equivalent to '
          'running ``sky launch -d ...`` and then ``sky autostop -i <minutes>``'
          '. If not set, the cluster will not be auto-stopped.'))
```

michaelzhiluo commented on sky/cli.py:
```diff
@@ -681,6 +704,9 @@ def _create_and_ssh_into_node(
         no_confirm: If true, skips confirmation prompt presented to user.
+        idle_minutes_to_autostop: Automatically stop the cluster after this
+                                  many minutes of idleness.
+        retry_until_up: Retry until up when true.
```
Improve description, look above.

ewzeng replied:
Sounds good! Should I put a condensed description here, or make them just as long as above.

michaelzhiluo replied:
Probably a summarized description of the above comment.

michaelzhiluo commented on sky/cli.py:
```diff
@@ -681,6 +704,9 @@ def _create_and_ssh_into_node(
         user_requested_resources: If true, user requested resources explicitly.
         no_confirm: If true, skips confirmation prompt presented to user.
+        idle_minutes_to_autostop: Automatically stop the cluster after this
+                                  many minutes of idleness.
```
`after specified minutes of idleness.`

ewzeng commented:
I updated the comments and the help messages.

I also added the `--zone` flag (didn't notice that #1205 requested it). Tested: `sky {cpunode, gpunode}` on GCP with varying regions and zones.

michaelzhiluo commented on sky/cli.py:
```diff
@@ -207,15 +207,53 @@ def _interactive_node_cli_command(cli_func):
+                                        'if we fail to launch the cluster on '
+                                        'any possible region/cloud due to '
+                                        'unavailability errors.'))
+    region_option = click.option('--region',
```
Small nit, group `retry_until_up` and `idle_autostop` and group `region_option` and `zone_option`

michaelzhiluo reviewed (approved):
LGTM, thanks @ewzeng ! Just need to address one small nit and I think its good to go 👍

ewzeng commented:
@michaelzhiluo How does it look now?

(Also I currently don't have write access to the repo, so you might need to help me merge)

michaelzhiluo commented:
LGTM, i'll merge in the next hour.

concretevitamin commented:
@ewzeng  A few potential issues:

1. We need to consider `region` and `zone` as specifying resources: https://github.com/skypilot-org/skypilot/blob/master/sky/cli.py#L2126-L2127.  This is not absolutely necessary (because `sky cpunode --region xx` will fail and hint users to do `sky cpunode --region xx --cloud yy`, where `--cloud` is indeed considered as specifying resources), but we should do it anyway.

2. I tried `sky cpunode` and then `sky cpunode -i 1`. The latter directly logs into the node and doesn't  set the autostop.  Can we add a smoke test?

ewzeng commented:
Thanks for the comments @concretevitamin!

1. What is the difference between placing `region` and `zone` in `user_requested_resources` vs. `Sky.resources`?
2. Didn't think about this situation. I presume the desired behavior is for autostop to be set?

concretevitamin commented:
1. To clarify, `user_requested_resources` is a bool, that's supposed to trigger https://github.com/skypilot-org/skypilot/blob/591ed685dd33caf7201f2e862fd60a6406dcfdb4/sky/cli.py#L757-L761. You could try this out by running `sky cpunode --cloud gcp -y` and then repeat this command again.  This comment is not meant to suggest moving things to `sky.Resources`.
2.  Yes, the desired behavior is to set the autostop.  This is a more pressing issue to be fixed -- if we can fix this by tomorrow's release, that'd be great; otherwise I filed #1220 to revert this PR for now.

ewzeng commented:
When is tomorrow's release? I will try to get #2 done, but not sure.
